tracer/src/Datadog.Trace/ClrProfiler/AutoInstrumentation/AWS/Kinesis/ContextPropagation.cs

```diff
@@ -0,0 +1,104 @@
+// <copyright file="ContextPropagation.cs" company="Datadog">
+// Unless explicitly stated otherwise all files in this repository are licensed under the Apache 2 License.
+// This product includes software developed at Datadog (https://www.datadoghq.com/). Copyright 2017 Datadog, Inc.
+// </copyright>
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using Datadog.Trace.DuckTyping;
+using Datadog.Trace.Logging;
+using Datadog.Trace.Propagators;
+using Datadog.Trace.Vendors.Newtonsoft.Json;
+
+namespace Datadog.Trace.ClrProfiler.AutoInstrumentation.AWS.Kinesis
+{
+    internal static class ContextPropagation
+    {
+        private const string KinesisKey = "_datadog";
+        private static readonly IDatadogLogger Log = DatadogLogging.GetLoggerFor(typeof(ContextPropagation));
+
+        public static void InjectTraceIntoRecords<TRecordsRequest>(IPutRecordsRequest request, SpanContext context)
+        {
+            // request.Records is not null and has at least one element
+            if (request.Records is not { Count: > 0 })
+            {
+                return;
+            }
+
+            var record = request.Records[0].DuckCast<IContainsData>();
+            InjectTraceIntoData<TRecordsRequest>(record, context);
```

Comment:
I don't know much about Kinesis, but do we need to inject the trace context into all the records? We're only putting it into the first one right now which seems strange to me at first glance

Reply:
Not really. Since distributed tracing is highly used for AWS Lambda and the AWS SDK integrations, our Datadog Lambda Extension only extracts from the first field or only field where we could inject trace context. 

When AWS Lambdas are triggered, they receive all the Records in the same payload.

We currently are not doing distributed tracing when using `GetRecords`, and will not be doing it until a customer requests it (might not happen). In that case, we'd have to change it to inject it on every Record, until then, this will suffice and will work perfectly for AWS Lambda.

Reply:
On an unrelated note, I don't think the generic call is correct right now. You can call it without the generic type argument and the call will still work with the type being inferred from the object, and the type will correctly correspond to the individual record type:

```suggestion
            InjectTraceIntoData(record, context);
```

Reply:
Thanks for the suggestion, updated the typings for Injection